### PR TITLE
moz_kinto_publisher: add MockClient for end-to-end testing

### PR DESCRIPTION
**Pull request by jschanck**

### CODE REVIEW

File: workflow/__init__.py [hunk 1]
```diff
@@ -15,7 +15,9 @@
 
 
 def get_test_dir(bucket_name):
-    return bucket_name.removeprefix(kTestBucket + ":")
+    if not bucket_name.startswith(kTestBucket+":"):
+        raise
```
Comment (mozkeeler) [inline workflow/__init__.py:19]: Should we raise a particular error?

**Review by mozkeeler — APPROVED** [review]
lgtm - just a couple of questions. Also, looks like the formatter is complaining in ci.

File: moz_kinto_publisher/test_publisher.py [hunk 2]
```diff
@@ -4,78 +4,124 @@
...
-            os.mkdir(Path(run) / mlbf_dir)
-            (Path(run) / mlbf_dir / "filter").touch()
-            (Path(run) / mlbf_dir / "filter.stash").touch()
-            (Path(run) / "ct-logs.json").touch()
-            (Path(run) / "completed").touch()
-            (Path(run) / "timestamp").write_text(
-                timestamp_from_run_id(run).isoformat(timespec="seconds"),
-                encoding="utf-8",
-            )
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (jschanck) [reply moz_kinto_publisher/test_publisher.py:20]: `self.run_identifiers` is initialized in the `__init__` of the superclass (`PublishedRunDB` in main.py). The superclass assumes that new runs will not be published during its lifetime, which is reasonable, but I wanted run_identifiers to be dynamic for tests.

File: moz_kinto_publisher/test_publisher.py [hunk 3]
```diff
@@ -4,78 +4,124 @@
...
+        record = kwargs["data"]
+        record["id"] = hashlib.sha256(
+            record["channel"].encode("utf-8")
+            + record["details"]["name"].encode("utf-8")
+        ).hexdigest()
+        self.existing_records += [record]
+        return {"data": record}
+
+    def delete_record(self, *args, **kwargs):
+        self.existing_records[:] = [
```
[the remark this answers is not in the document]
Comment (jschanck) [reply moz_kinto_publisher/test_publisher.py:99]: not super important here, but slice assignment changes the object referenced by `self.existing_records` whereas a regular assignment would bind `self.existing_records` to a new object.

File: workflow/__init__.py (hunk 1 above)
Comment (jschanck) [reply workflow/__init__.py]: done!

**Pull request merged**